### ISSUE TITLE
RFC: Crazy idea: rewrite frontend in React?

### ISSUE THREAD

**Issue by danielbeardsley** [iFixit/pulldasher#199]
In progress: https://github.com/iFixit/pulldasher/compare/reaaaaaaaaaaact

### Problem

The somewhat confusing mix of old-style JS (inspecting + mutating the dom), js templates, events, and the super-configurable design of the spec of the frontend mean that adding new features is harder than it could be if the JS design and organization were more familiar / typical.

### Suggestion
Re-create the *front-end* entirely as an SPA in React. To be clear, the front end is basically a visualizer around an array of json `pull` objects that occasionally get's updated from the backend.

#### Pros
* UI is a perfect candidate for React
  * Components are well defined (page, column, cards)
  * Components can be tested and mocked in isolation
  * State updates come from one place and just amount to `setState(newPullsState)`
  * Other bits on the page are easy to imagine as react components
  * Front-end can *easily* be stubbed with example data for local testing
* As a team we'd feel a lot more comfortable maintaining the front-end
  * There's quite a bit of deferred fixes / adjustments that are hard to make cause it's hard to comprehend
  * Also cause it's hard to test in isolation. We shouldn't need a MySQL DB to test the front-end
* So fun!

#### Cons
* It works as is
* Decent amount of effort (2 or 3 days)
* ???

CC @andyg0808 @jarstelfox @djmetzle @sctice @sterlinghirsh @davidrans ....

**Comment by andyg0808** [comment]
The front-end already rerenders on data change. It's just begging to be turned into React. +100 from me!

Might be worth a few minutes consideration whether the frontend data should also be moved to Redux; I'm suspicious the answer is yes.

**Comment by danielbeardsley** [comment]
Also, these two frontends could live in parallel while one is approaching feature-parity, thus not making us keep a giant un-merged pull around.

**Comment by jarstelfox** [comment]
I say go for it. It will allow us to easily add features.

**Comment by danielbeardsley** [comment]
I stayed up late and played with this. We have a typescript / react starting point.

https://github.com/iFixit/pulldasher/compare/reaaaaaaaaaaact

**Comment by andyg0808** [comment]
+1 for choice of branch name.

**Comment by andyg0808** [comment]
Might be worth adding https://www.typescriptlang.org/tsconfig#noImplicitAny to the config, for reduced frustration, but looks nice so far!

**Comment by danielbeardsley** [comment]
Anyone is welcome to pick up this torch and carry on!

I actually got the sending-pulls-to-the-frontend part working, but I backed off and did typescript first.

The requirements are:

* Be able to easily stub the backend
   * An example json file (of pulls) that is easy to use without having to set anything up in github nor run mysql
   * A way of pushing fake pull updates
      * Maybe have a special debug mode where the refresh endpoint just randomly mutates a pull instead of sourcing it from the API
   * Include enough data that most UI situations are shown
   * npm script to run the backend in "fake-mode"

**Comment by danielbeardsley** [comment]
FYI, there's an open pull!

**Comment by danielbeardsley** [comment]
Some code is merged, prgress is slowly being made. This RFC is done.
